src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+julia> get(d, "c", 3)
+3
+```
+"""
+get(collection, key, default)
+
+function get(h::RobinDict{K,V}, key0, default) where {K, V}
+    key = convert(K, key0)
+    index = rh_search(h, key)
+    @inbounds return (index < 0) ? default : h.vals[index]::V
+end
+
+"""
+    get(f::Function, collection, key)
+
+Return the value stored for the given key, or if no mapping for the key is present, return
+`f()`.  Use [`get!`](@ref) to also store the default value in the dictionary.
+
+This is intended to be called using `do` block syntax
+
+```julia
+get(dict, key) do
+    # default value calculated here
+    time()
+end
+```
+"""
+get(::Function, collection, key)
+
+function get(default::Callable, h::RobinDict{K,V}, key0) where {K, V}
+    key = convert(K, key0) 
+    index = rh_search(h, key)
+    @inbounds return (index < 0) ? default() : h.vals[index]::V
+end
+
+"""
+    haskey(collection, key) -> Bool
+
+Determine whether a collection has a mapping for a given `key`.
+
+# Examples
+```jldoctest
+julia> D = RobinDict('a'=>2, 'b'=>3)
+RobinDict{Char,Int64} with 2 entries:
+  'a' => 2
+  'b' => 3
+
+julia> haskey(D, 'a')
+true
+
+julia> haskey(D, 'c')
+false
+```
+"""
+haskey(h::RobinDict, key) = (rh_search(h, key) > 0) 
+in(key, v::KeySet{<:Any, <:RobinDict}) = (rh_search(v.dict, key) >= 0)
+
+"""
+    getkey(collection, key, default)
+
+Return the key matching argument `key` if one exists in `collection`, otherwise return `default`.
+
+# Examples
+```jldoctest
+julia> D = RobinDict('a'=>2, 'b'=>3)
+RobinDict{Char,Int64} with 2 entries:
+  'a' => 2
+  'b' => 3
+
+julia> getkey(D, 'a', 1)
+'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
+
+julia> getkey(D, 'd', 'a')
+'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
+```
+"""
+function getkey(h::RobinDict{K,V}, key0, default) where {K, V}
+    key = convert(K, key0) 
+    index = rh_search(h, key)
+    @inbounds return (index < 0) ? default : h.keys[index]::K
+end
+
+# backward shift deletion by not keeping any tombstones
+function rh_delete!(h::RobinDict{K, V}, index) where {K, V}
+    @assert index > 0
+
+    # this assumes that there is a key/value present in the dictionary at index
+    index0 = index
+    sz = length(h.keys)
+    @inbounds while true
+        index0 = (index0 & (sz - 1)) + 1
+        if isslotempty(h, index0) || calculate_distance(h, index0) == 0
+            break
+        end
+    end
+    #index0 represents the position before which we have to shift backwards 
+    
+    # the backwards shifting algorithm
+    curr = index
+    next = (index & (sz - 1)) + 1
+    
+    @inbounds while next != index0
+        h.vals[curr] = h.vals[next]
+        h.keys[curr] = h.keys[next]
+        h.hashes[curr] = h.hashes[next]
+        curr = next
+        next = (next & (sz-1)) + 1
+    end
+    
+    #curr is at the last position, reset back to normal
+    ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.keys, curr-1)
+    ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.vals, curr-1)
```

Comment:
What is thsi about?
Why do we need to unset elements in an array?
Are we later checking them with `isdefined`?

Reply:
This way the way deletions were happening in `base/dict.jl`, so I kept this part of the code. No, aren't checking for `isdefined`, checking whether `hashes` is `0` or not, seems sufficient

Reply:
@vtjnash maybe can comment.
If there is a good reason to do this we should explain why,
if not we should just not do this.

Reply:
We need to unset the array indices for key/value types that are stored by garbage collected reference. Otherwise, the garbage collector will see the reference and keep the object alive. 

The unset-call is unneeded for bitstypes and bits unions; it should look exactly like Base here:
```
isbitstype(K) || isbitsunion(K) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.keys, curr-1)
isbitstype(V) || isbitsunion(V) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.vals, curr-1)
```

Reply:
Correct. The checks are non-essential (jl_arrayunset must execute them also), but generally it's good to match Base anyways.